mod/file/start.php

```diff
@@ -359,66 +359,86 @@ function file_set_url($hook, $type, $url, $params) {
 function file_set_icon_url($hook, $type, $url, $params) {
 	$file = $params['entity'];
 	$size = $params['size'];
-	if (elgg_instanceof($file, 'object', 'file')) {
+	if (!elgg_instanceof($file, 'object', 'file')) {
+		return;
+	}
 
-		// thumbnails get first priority
-		if ($file->thumbnail) {
-			$ts = (int)$file->icontime;
-			return "mod/file/thumbnail.php?file_guid=$file->guid&size=$size&icontime=$ts";
+	// thumbnails get first priority
+	if ($file->thumbnail) {
+		
+		switch ($size) {
+			case "small":
+				$thumbfile = $file->thumbnail;
+				break;
+			case "medium":
+				$thumbfile = $file->smallthumb;
+				break;
+			case "large":
+			default:
+				$thumbfile = $file->largethumb;
+				break;
 		}
 
-		$mapping = array(
-			'application/excel' => 'excel',
-			'application/msword' => 'word',
-			'application/ogg' => 'music',
-			'application/pdf' => 'pdf',
-			'application/powerpoint' => 'ppt',
-			'application/vnd.ms-excel' => 'excel',
-			'application/vnd.ms-powerpoint' => 'ppt',
-			'application/vnd.oasis.opendocument.text' => 'openoffice',
-			'application/vnd.openxmlformats-officedocument.wordprocessingml.document' => 'word',
-			'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' => 'excel',
-			'application/vnd.openxmlformats-officedocument.presentationml.presentation' => 'ppt',
-			'application/x-gzip' => 'archive',
-			'application/x-rar-compressed' => 'archive',
-			'application/x-stuffit' => 'archive',
-			'application/zip' => 'archive',
-
-			'text/directory' => 'vcard',
-			'text/v-card' => 'vcard',
-
-			'application' => 'application',
-			'audio' => 'music',
-			'text' => 'text',
-			'video' => 'video',
-		);
-
-		$mime = $file->mimetype;
-		if ($mime) {
-			$base_type = substr($mime, 0, strpos($mime, '/'));
-		} else {
-			$mime = 'none';
-			$base_type = 'none';
+		if (!empty($thumbfile)) {
+			$readfile = new ElggFile();
+			$readfile->owner_guid = $file->owner_guid;
+			$readfile->setFilename($thumbfile);
+			$thumb_url = elgg_get_inline_url($readfile, true);
+			if ($thumb_url) {
+				return $thumb_url;
```

Comment:
We don't have an icon for simpletype = image. Need to fallback to default entity icon?